File: poly/DesignPatterns/Factory/factory.hs
-- Factory design pattern
import Data.Char  -- toUpper

class IShape a where
  draw :: a -> IO ()

data Rectangle = Rectangle
instance IShape Rectangle where
  draw Rectangle = putStrLn "Inside Rectangle::draw() method."

data Square = Square
instance IShape Square where
  draw Square = putStrLn "Inside Square::draw() method."

data Circle = Circle
instance IShape Circle where
  draw Circle = putStrLn "Inside Circle::draw() method."

data Shape = MkShapeR Rectangle | MkShapeS Square | MkShapeC Circle
instance IShape Shape where
  draw (MkShapeR x) = draw x
  draw (MkShapeS x) = draw x
  draw (MkShapeC x) = draw x

data ShapeFactory = ShapeFactory
getShape :: ShapeFactory -> String -> Maybe Shape
getShape ShapeFactory shapeType = case (map toUpper shapeType) of
   "CIRCLE"    -> Just (MkShapeC Circle)
   "RECTANGLE" -> Just (MkShapeR Rectangle)
   "SQUARE"    -> Just (MkShapeS Square)
   otherwise   -> Nothing

main = do 
  let shapeFactory = ShapeFactory
      Just shape1 = getShape shapeFactory "CIRCLE"
      Just shape2 = getShape shapeFactory "RECTANGLE"
      Just shape3 = getShape shapeFactory "SQUARE"
    in do
      draw shape1
      draw shape2
      draw shape3

File: poly/DesignPatterns/Factory/factory.py
from abc import ABCMeta
from abc import abstractmethod

class IShape(metaclass=ABCMeta):

    @abstractmethod
    def draw(self):
        return NotImplemented

class Rectangle(IShape):

    def draw(self):
        print("Inside Rectangle::draw() method.")

class Square(IShape):

    def draw(self):
        print("Inside Square::draw() method.")

class Circle(IShape):

    def draw(self):
        print("Inside Circle::draw() method.")


class ShapeFactory(object):
    # use getShape method to get object of type IShape
    def getShape(self, shapeType):

        if not shapeType:   # testing for empty string
            return None

        if shapeType.upper() == "CIRCLE":
            return Circle()
        elif shapeType.upper() == "RECTANGLE":
            return Rectangle()
        elif shapeType.upper() == "SQUARE":
            return Square()

        return None


shapeFactory = ShapeFactory()
# get an object of Circle and call its draw method
shape1 = shapeFactory.getShape("CIRCLE")
shape1.draw()

# get an object of Rectangle and call its draw method
shape2 = shapeFactory.getShape("RECTANGLE")
shape2.draw()

# get an object of Square and call its draw method
shape3 = shapeFactory.getShape("SQUARE")
shape3.draw()

